### PR TITLE
Prevent crash when a service is added or removed during handle_response

### LINKED ISSUE #316: self.services can change during iteration in `handle_query`
```
2020-10-25 18:57:15 ERROR (zeroconf-Engine-257) [root] Uncaught thread exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.8/site-packages/zeroconf/__init__.py", line 1355, in run
    reader.handle_read(socket_)
  File "/usr/local/lib/python3.8/site-packages/zeroconf/__init__.py", line 1448, in handle_read
    self.zc.handle_query(msg, None, _MDNS_PORT)
  File "/usr/local/lib/python3.8/site-packages/zeroconf/__init__.py", line 2780, in handle_query
    for service in self.services.values():
RuntimeError: dictionary changed size during iteration
```

This ends up being an uncaught exception as well so `zeroconf` dies off.

I'm opening the issue so I don't forget to look into it.

### PR DESCRIPTION
**Pull request by bdraco**
Fixes #316

### CODE REVIEW

File: zeroconf/__init__.py [hunk 1]
```diff
@@ -2766,69 +2767,77 @@ def handle_query(self, msg: DNSIncoming, addr: Optional[str], port: int) -> None
...
                             )
-                        )
-                        out.add_additional_answer(
-                            DNSText(
-                                service.name,
-                                _TYPE_TXT,
-                                _CLASS_IN | _CLASS_UNIQUE,
-                                service.other_ttl,
-                                service.text,
+                    for service in self.services.values():
```
Comment (bdraco) [inline zeroconf/__init__.py:2786]: I could modify `self.servicetypes` under the service lock to make it a `dict` of types with the values as a list of `ServiceInfo` so we can do 

`self.servicetypes.get(question.name)` and then copy the list so we only copy on a hit.

File: zeroconf/__init__.py [hunk 2]
```diff
@@ -2766,69 +2767,77 @@ def handle_query(self, msg: DNSIncoming, addr: Optional[str], port: int) -> None
...
+                                )
+                else:
                     if out is None:
                         out = DNSOutgoing(_FLAGS_QR_RESPONSE | _FLAGS_AA)
 
+                    name_to_find = question.name.lower()
+
                     # Answer A record queries for any service addresses we know
                     if question.type in (_TYPE_A, _TYPE_ANY):
                         for service in self.services.values():
```
Comment (bdraco) [inline zeroconf/__init__.py:2839]: Here I could add `self.serviceservers` under the service lock to make it a `dict` of types with the values as a list of `ServiceInfo` so we can do 

`self.serviceservers.get(name_to_find)` and then copy the list so we only copy on a hit.

Comment (bdraco) [reply zeroconf/__init__.py:2839]: I'd need to pop it back out of both of them in `update_service` and put the new one in.

**Comment by bdraco** [comment]
This will prevent it from crashing, but it will still hit `RuntimeError` if a service is added or removed and will end up dropping the incoming packet.  That's a lot better than a unhandled exception.

I can fix that but its a larger change as suggested above.  I'm looking for guidance on how to proceed.

**Comment by jstasiak** [comment]
Oh boy, this seems like a tricky one. I'll get back to you in a few days, I need to reread the relevant parts of the code when I have some time.

**Comment by bdraco** [comment]
I'm thinking it might be cleaner to pull out the service registration into a new class so we can abstract away the get/set/locking operations

**Comment by bdraco** [comment]
I've been running this for a while now on 5 systems with no issues and no more randomly failing startup

**Comment by bdraco** [comment]
Almost a month running this an my automation to restart Home Assistant when HomeKit fails to be seen on the network hasn't triggered once even though I've restarted hundreds of times for dev testing so I'm confident this is solid now.

**Comment by bdraco** [comment]
@jstasiak Any chance you have a moment to look this over soon?

**Comment by jstasiak** [comment]
I'm slightly hesitant regarding merging this as a single thing – are you fine with merging the first two commits (the actual fix) as one thing and then the rest (refactoring) as another?

File: zeroconf/__init__.py [hunk 3]
```diff
@@ -2930,23 +2963,24 @@ def send(self, out: DNSOutgoing, addr: Optional[str] = None, port: int = _MDNS_P
...
-            self._GLOBAL_DONE = True
-
-            # shutdown recv socket and thread
-            if not self.unicast:
-                self.engine.del_reader(cast(socket.socket, self._listen_socket))
-                cast(socket.socket, self._listen_socket).close()
-            for s in self._respond_sockets:
-                self.engine.del_reader(s)
-            self.engine.join()
+        if self._GLOBAL_DONE:
```
Comment (jstasiak) [inline zeroconf/__init__.py:2966]: Let's restore the previous indentation here, it makes the patch larger than needed.

Comment (bdraco) [reply zeroconf/__init__.py]: Reverted

**Comment by bdraco** [comment]
> I'm slightly hesitant regarding merging this as a single thing – are you fine with merging the first two commits (the actual fix) as one thing and then the rest (refactoring) as another?

I can split it/squash it however you would like it. There are a few more places in the second part that address other potential thread safety points were the refactoring was needed to move the changes under the lock (`for stype in self.servicetypes.keys` and `for service in self.services.values` being removed)

**Comment by jstasiak** [comment]
> How about I squash the first part into this PR, and then open a second PR with the squashed service registry?

This is exactly what I had in mind, would make reviewing this now and going through the code later that much easier!

**Comment by bdraco** [comment]
I've forced pushed this one with the additional commits removed, and squashed them into another PR here https://github.com/jstasiak/python-zeroconf/pull/320

**Comment by jstasiak** [comment]
Lovely, thanks!

**Pull request merged**